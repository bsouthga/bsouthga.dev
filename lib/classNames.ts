export default function classNames(
  ...names: (string | null | undefined)[]
): string {
  return names.filter(Boolean).join(" ");
}
