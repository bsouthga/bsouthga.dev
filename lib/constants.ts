export const TYPESTYLE_TARGET = "typestyle-target";
