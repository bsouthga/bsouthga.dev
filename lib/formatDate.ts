const OPTIONS: Intl.DateTimeFormatOptions = {
  year: "numeric",
  month: "long",
  day: "numeric",
};

export default function formatDate(datestr: string) {
  const date = new Date(datestr);
  return date.toLocaleDateString(undefined, OPTIONS);
}
