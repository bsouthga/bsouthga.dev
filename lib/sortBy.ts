export default function sortBy<T>(arr: T[], valueFn: (v: T) => number): T[] {
  const values = arr.slice();

  values.sort((a, b) => {
    const aV = valueFn(a);
    const bV = valueFn(b);
    return bV - aV;
  });

  return values;
}
