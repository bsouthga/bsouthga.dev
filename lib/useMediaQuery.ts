import { useEffect, useState } from "react";

function getMediaQueryList(query: string): MediaQueryList | null {
  if (typeof window === "undefined" || window.matchMedia == null) {
    return null;
  }

  return window.matchMedia(query);
}

function matchesMedia(query: string): boolean {
  return getMediaQueryList(query)?.matches ?? false;
}

export default function useMediaQuery(query: string): boolean {
  const [value, setValue] = useState(matchesMedia(query));

  useEffect(() => {
    const list = getMediaQueryList(query);
    const handler = () => setValue(matchesMedia(query));
    list.addEventListener("change", handler);
    return () => list.removeEventListener("change", handler);
  }, [query]);

  return value;
}
