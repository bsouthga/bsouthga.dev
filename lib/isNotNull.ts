import { Maybe } from "lib/types";

export default function isNotNull<T>(value: Maybe<T>): value is T {
  return value != null;
}
