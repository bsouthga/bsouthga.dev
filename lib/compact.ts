import { Maybe } from "lib/types";
import isNotNull from "lib/isNotNull";

export default function compact<T>(values: Maybe<T>[]): T[] {
  return values.filter(isNotNull);
}
