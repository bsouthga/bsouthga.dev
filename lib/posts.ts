import fs from "fs";
import path from "path";
import getConfig from "next/config";
import sortBy from "lib/sortBy";
import { Maybe } from "lib/types";
import compact from "lib/compact";

const { serverRuntimeConfig } = getConfig() ?? {
  serverRuntimeConfig: {
    PROJECT_ROOT: process.cwd(),
  },
};
const POST_DIR = path.join(
  serverRuntimeConfig.PROJECT_ROOT,
  "./public/markdown/"
);

export type PostMetadata = {
  date: string;
  subtitle: string;
  title: string;
  filename: string;
  content: string;
  github?: Maybe<string>;
  metaImage?: Maybe<string>;
};

export async function getPostFiles(): Promise<string[]> {
  const files = await fs.promises.readdir(POST_DIR);
  return files.map((filename) => filename.replace(".md", ""));
}

export async function getPost(id: string): Promise<Maybe<PostMetadata>> {
  const file = (await fs.promises.readFile(`${POST_DIR}/${id}.md`)).toString();
  return parsePostMetadata(id, file);
}

export async function getAllPosts(): Promise<PostMetadata[]> {
  const files = await getPostFiles();
  const posts = await Promise.all(files.map(getPost));
  return sortBy(compact(posts), (post) => new Date(post.date).getTime());
}

function parsePostMetadata(
  filename: string,
  post: string
): Maybe<PostMetadata> {
  const [__, metadataRaw, content] = post.split("---");
  const metadata = metadataRaw.split("\n").reduce((out, line) => {
    const [key, ...values] = line.split(":");
    out[key.replace(/\W/g, "").trim()] = values.join(":").trim();
    return out;
  }, {} as { [key: string]: Maybe<string> });

  const { date, subtitle, title, metaImage, github } = metadata;
  if (date == null || subtitle == null || title == null) {
    return null;
  }

  return {
    date: date,
    subtitle: subtitle,
    title: title,
    metaImage: metaImage ?? null,
    github: github ?? null,
    filename: filename.replace(/\.md$/, ""),
    content,
  };
}
