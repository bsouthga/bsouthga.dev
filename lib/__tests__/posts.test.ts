import { getPost } from "lib/posts";

const POSTS = ["pca-image-classification", "color-gradients-with-python"];

describe("posts", () => {
  test.each(POSTS)("snapshot metadata", async (postID) => {
    const post = await getPost(postID);
    expect(post).toBeTruthy();
    const { content, ...rest } = post!;
    expect(content).toBeTruthy();
    expect(rest).toMatchSnapshot(postID);
  });
});
