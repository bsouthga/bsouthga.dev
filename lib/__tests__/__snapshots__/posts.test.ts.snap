// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`posts snapshot metadata: color-gradients-with-python 1`] = `
Object {
  "date": "12/03/2013",
  "filename": "color-gradients-with-python",
  "github": null,
  "metaImage": "/assets/images/bezier_example.png",
  "subtitle": "Investigating a few \\"Pythonic\\" ways to calculate gradients across given input colors",
  "title": "Color gradients with Python",
}
`;

exports[`posts snapshot metadata: pca-image-classification 1`] = `
Object {
  "date": "08/05/2017",
  "filename": "pca-image-classification",
  "github": "https://github.com/bsouthga/pca-image-classification",
  "metaImage": "/assets/images/eig_1_heatmap.png",
  "subtitle": "Project for STAT6215 at GWU (part of my M.S. coursework) using classic statistical methods (Principal Component and Linear Discriminant Analysis) to classify images.",
  "title": "Image Classification using PCA and LDA",
}
`;
